sql/analyzer/indexed_joins.go

```diff
@@ -819,13 +819,13 @@ func getJoinIndex(
 	for table, cols := range exprsByTable {
 		exprs := extractExpressions(cols)
 		idx := ia.MatchingIndex(ctx, ctx.GetCurrentDatabase(), table, normalizeExpressions(ctx, tableAliases, exprs...)...)
-		// If we do not find a perfect index, we take the first single column partial index if there is one.
-		// This currently only finds single column indexes. A better search would look for the most complete
-		// index available, covering the columns with the most specificity / highest cardinality.
+		// If we do not find a full or partial matching index, we take the first single column index if there is one.
```

Comment:
This is fundamental enough that we need engine tests for this as well, not just a single script test.

Like why doesn't this engine test catch the bug?

```
"SELECT opk.c5,pk1,pk2 FROM one_pk opk JOIN two_pk tpk ON opk.pk=tpk.pk1 ORDER BY 1,2,3"
```

I'm also not a huge fan of this interface pattern, where `ia.MatchingIndex` gets called with various permutations in a loop. `MatchingIndex` should have logic to pick the best index, *or* to return all possible matches in some consistent ordering. That can be a TODO I guess.

Reply:
`MatchingIndex` is a rename of the old function `IndexByExpression`

https://github.com/dolthub/go-mysql-server/blob/51cfa9bbc386a1508a8804114ff0a5a034a5e096/sql/analyzer/indexed_joins.go#L816

I didn't modify/change anything here in the index interface rewrite besides updating the function name and now letting it work better with partial matching and non-prefixed columns.

Reply:
`MatchingIndex` does pick the best index, in that it returns the most exact match first. That's code is just doing some extra stuff around that.